std/container/array.d

```diff
@@ -59,87 +48,255 @@ if (!is(Unqual!T == bool))
         //     return result;
         // }
 
-        // length
+        // Returns the number of elements in the Array.
         @property size_t length() const
         {
             return _payload.length;
         }
 
-        // length
+        // Sets the number of elements in the Array.
+        // The length of the Array can never be more than the capacity.
+        // Note: Reducing the length will remove items
+        // from the end of the Array.
         @property void length(size_t newLength)
         {
-            if (length >= newLength)
-            {
-                // shorten
-                static if (hasElaborateDestructor!T)
-                    foreach (ref e; _payload.ptr[newLength .. _payload.length])
-                        .destroy(e);
-
-                _payload = _payload.ptr[0 .. newLength];
-                return;
-            }
-            // enlarge
-            auto startEmplace = length;
-            _payload = (cast(T*) realloc(_payload.ptr,
-                            T.sizeof * newLength))[0 .. newLength];
-            initializeAll(_payload.ptr[startEmplace .. length]);
+            if (newLength > capacity)
+                capacity = newLength; // enlarge
+            if (newLength < length)
+                removeAt(newLength, length - newLength); // shorten
+            _payload = _payload.ptr[0 .. newLength];
         }
 
-        // capacity
+        // Returns the capacity of the Array.
         @property size_t capacity() const
         {
             return _capacity;
         }
 
-        // reserve
-        void reserve(size_t elements)
+        // Set the capacity of the Array.
+        // The capacity is the maximum size of the Array before it
+        // must be resized.
+        // The capacity cannot be less than the Array length.
+        // Setting capacity to 0 will free the Array any other value
+        // will cause a resize.
+        @property void capacity(size_t value)
```

Comment:
The capacity and length can both be shrank. If you shrink the capacity < length it will remove the elements accordingly, before it de-allocates them.